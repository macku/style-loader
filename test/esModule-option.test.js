import path from "path";

import {
  compile,
  getCompiler,
  getEntryByInjectType,
  getErrors,
  getWarnings,
  runInJsDom,
} from "./helpers/index";

describe('"esModule" option', () => {
  const injectTypes = [
    "styleTag",
    "singletonStyleTag",
    "lazyStyleTag",
    "lazySingletonStyleTag",
    "linkTag",
  ];

  const moduleTypes = ['commonjs-modules.js', 'es-modules.js'];

  const esModuleValues = [
    // No CSS modules  
    { styleLoader: {}, cssLoader: {} },
    { styleLoader: { esModules: false }, cssLoader: { esModules: false } },
    { styleLoader: { esModules: true }, cssLoader: { esModules: false } },
    { styleLoader: { esModules: false }, cssLoader: { esModules: true } },
    { styleLoader: { esModules: true }, cssLoader: { esModules: true } },
    { styleLoader: {}, cssLoader: { modules: false } },
      
    // CSS modules
    { styleLoader: {}, cssLoader: { modules: true } },
    { styleLoader: {}, cssLoader: { modules: { namedExport: false } } },
    { styleLoader: {}, cssLoader: { modules: { namedExport: true } } },
    { styleLoader: { esModules: false }, cssLoader: { esModules: false, modules: { namedExport: false } } },
    { styleLoader: { esModules: true }, cssLoader: { esModules: false, modules: { namedExport: false } } },
    { styleLoader: { esModules: false }, cssLoader: { esModules: true, modules: { namedExport: false } } },
    { styleLoader: { esModules: true }, cssLoader: { esModules: true, modules: { namedExport: false } } },
    { styleLoader: { esModules: false }, cssLoader: { esModules: false, modules: { namedExport: true } } },
    { styleLoader: { esModules: true }, cssLoader: { esModules: false, modules: { namedExport: true } } },
    { styleLoader: { esModules: false }, cssLoader: { esModules: true, modules: { namedExport: true } } },
    { styleLoader: { esModules: true }, cssLoader: { esModules: true, modules: { namedExport: true } } },
  ];

  injectTypes.forEach((injectType) => {
    esModuleValues.forEach((esModuleValue) => {
      const isNamedExport =
        esModuleValue === "named" && injectType !== "linkTag";

      let testName = "should work";
      testName += `${
        esModuleValue === "not specified"
          ? " when not specified"
          : isNamedExport
          ? " when modules enabled"
          : esModuleValue === true
          ? ' with a value equal to "true"'
          : esModuleValue === false
          ? ` with a value equal to "false"`
          : ""
      }`;
      testName += ` and when the "injectType" option is "${injectType}"`;

      it(`${testName} and ES module syntax used`, async () => {
        const entry = getEntryByInjectType("es-modules.js", injectType);
        const compiler = getCompiler(
          entry,
          { injectType },
          {
            module: {
              rules: [
                {
                  test: /\.css$/i,
                  use: [
                    {
                      loader: path.resolve(__dirname, "../src/cjs.js"),
                      options: esModuleValue.styleLoader,
                    },
                    injectType === "linkTag"
                      ? {
                          loader: "file-loader",
                          options: {
                            esModule: true,
                            name: "[path][name].[ext]",
                          },
                        }
                      : {
                          loader: "css-loader",
                          options: esModuleValue.cssLoader,
                        },
                  ],
                },
              ],
            },
          }
        );
        const stats = await compile(compiler);

        runInJsDom("main.bundle.js", compiler, stats, (dom) => {
          expect(dom.serialize()).toMatchSnapshot("DOM");

          // TODO test using global variable
          // expect(exports).toMatchSnapshot("exports");
        });

        expect(getWarnings(stats)).toMatchSnapshot("warnings");
        expect(getErrors(stats)).toMatchSnapshot("errors");
      });
    });
  });
});
