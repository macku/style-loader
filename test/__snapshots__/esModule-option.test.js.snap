// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"esModule" option should work and when the "injectType" option is "linkTag": DOM 1`] = `
"<!DOCTYPE html><html><head>
  <title>style-loader test</title>
  <style id=\\"existing-style\\">.existing { color: yellow }</style>
<link rel=\\"stylesheet\\" href=\\"es-modules.css\\"></head>
<body>
  <h1>Body</h1>
  <div class=\\"target\\"></div>
  <iframe class=\\"iframeTarget\\"></iframe>


<div>
============================
EXPORTS:
============================
{
	\\"default\\": {}
}
============================
</div></body></html>"
`;

exports[`"esModule" option should work and when the "injectType" option is "linkTag": errors 1`] = `Array []`;

exports[`"esModule" option should work and when the "injectType" option is "linkTag": warnings 1`] = `Array []`;

exports[`"esModule" option should work when modules enabled and when the "injectType" option is "lazySingletonStyleTag": DOM 1`] = `
"<!DOCTYPE html><html><head>
  <title>style-loader test</title>
  <style id=\\"existing-style\\">.existing { color: yellow }</style>
<style>.foo {
    color: red;
}

.bar {
    color: blue;
}

.baz {
    color: coral;
}
</style></head>
<body>
  <h1>Body</h1>
  <div class=\\"target\\"></div>
  <iframe class=\\"iframeTarget\\"></iframe>


<div>
EXPORTED defaultImport:
============================
{}
============================
EXPORTED allImports:
============================
{
	\\"default\\": {}
}
============================
</div></body></html>"
`;

exports[`"esModule" option should work when modules enabled and when the "injectType" option is "lazySingletonStyleTag": errors 1`] = `Array []`;

exports[`"esModule" option should work when modules enabled and when the "injectType" option is "lazySingletonStyleTag": warnings 1`] = `Array []`;

exports[`"esModule" option should work when modules enabled and when the "injectType" option is "lazyStyleTag": DOM 1`] = `
"<!DOCTYPE html><html><head>
  <title>style-loader test</title>
  <style id=\\"existing-style\\">.existing { color: yellow }</style>
<style>.foo {
    color: red;
}

.bar {
    color: blue;
}

.baz {
    color: coral;
}
</style></head>
<body>
  <h1>Body</h1>
  <div class=\\"target\\"></div>
  <iframe class=\\"iframeTarget\\"></iframe>


<div>
EXPORTED defaultImport:
============================
{}
============================
EXPORTED allImports:
============================
{
	\\"default\\": {}
}
============================
</div></body></html>"
`;

exports[`"esModule" option should work when modules enabled and when the "injectType" option is "lazyStyleTag": errors 1`] = `Array []`;

exports[`"esModule" option should work when modules enabled and when the "injectType" option is "lazyStyleTag": warnings 1`] = `Array []`;

exports[`"esModule" option should work when modules enabled and when the "injectType" option is "singletonStyleTag": DOM 1`] = `
"<!DOCTYPE html><html><head>
  <title>style-loader test</title>
  <style id=\\"existing-style\\">.existing { color: yellow }</style>
<style>.foo {
    color: red;
}

.bar {
    color: blue;
}

.baz {
    color: coral;
}
</style></head>
<body>
  <h1>Body</h1>
  <div class=\\"target\\"></div>
  <iframe class=\\"iframeTarget\\"></iframe>


<div>
============================
EXPORTS:
============================
{
	\\"default\\": {}
}
============================
</div></body></html>"
`;

exports[`"esModule" option should work when modules enabled and when the "injectType" option is "singletonStyleTag": errors 1`] = `Array []`;

exports[`"esModule" option should work when modules enabled and when the "injectType" option is "singletonStyleTag": warnings 1`] = `Array []`;

exports[`"esModule" option should work when modules enabled and when the "injectType" option is "styleTag": DOM 1`] = `
"<!DOCTYPE html><html><head>
  <title>style-loader test</title>
  <style id=\\"existing-style\\">.existing { color: yellow }</style>
<style>.foo {
    color: red;
}

.bar {
    color: blue;
}

.baz {
    color: coral;
}
</style></head>
<body>
  <h1>Body</h1>
  <div class=\\"target\\"></div>
  <iframe class=\\"iframeTarget\\"></iframe>


<div>
============================
EXPORTS:
============================
{
	\\"default\\": {}
}
============================
</div></body></html>"
`;

exports[`"esModule" option should work when modules enabled and when the "injectType" option is "styleTag": errors 1`] = `Array []`;

exports[`"esModule" option should work when modules enabled and when the "injectType" option is "styleTag": warnings 1`] = `Array []`;

exports[`"esModule" option should work when not specified and when the "injectType" option is "lazySingletonStyleTag": DOM 1`] = `
"<!DOCTYPE html><html><head>
  <title>style-loader test</title>
  <style id=\\"existing-style\\">.existing { color: yellow }</style>
<style>.foo {
    color: red;
}

.bar {
    color: blue;
}

.baz {
    color: coral;
}
</style></head>
<body>
  <h1>Body</h1>
  <div class=\\"target\\"></div>
  <iframe class=\\"iframeTarget\\"></iframe>


<div>
EXPORTED defaultImport:
============================
{}
============================
EXPORTED allImports:
============================
{
	\\"default\\": {}
}
============================
</div></body></html>"
`;

exports[`"esModule" option should work when not specified and when the "injectType" option is "lazySingletonStyleTag": errors 1`] = `Array []`;

exports[`"esModule" option should work when not specified and when the "injectType" option is "lazySingletonStyleTag": warnings 1`] = `Array []`;

exports[`"esModule" option should work when not specified and when the "injectType" option is "lazyStyleTag": DOM 1`] = `
"<!DOCTYPE html><html><head>
  <title>style-loader test</title>
  <style id=\\"existing-style\\">.existing { color: yellow }</style>
<style>.foo {
    color: red;
}

.bar {
    color: blue;
}

.baz {
    color: coral;
}
</style></head>
<body>
  <h1>Body</h1>
  <div class=\\"target\\"></div>
  <iframe class=\\"iframeTarget\\"></iframe>


<div>
EXPORTED defaultImport:
============================
{}
============================
EXPORTED allImports:
============================
{
	\\"default\\": {}
}
============================
</div></body></html>"
`;

exports[`"esModule" option should work when not specified and when the "injectType" option is "lazyStyleTag": errors 1`] = `Array []`;

exports[`"esModule" option should work when not specified and when the "injectType" option is "lazyStyleTag": warnings 1`] = `Array []`;

exports[`"esModule" option should work when not specified and when the "injectType" option is "linkTag": DOM 1`] = `
"<!DOCTYPE html><html><head>
  <title>style-loader test</title>
  <style id=\\"existing-style\\">.existing { color: yellow }</style>
<link rel=\\"stylesheet\\" href=\\"es-modules.css\\"></head>
<body>
  <h1>Body</h1>
  <div class=\\"target\\"></div>
  <iframe class=\\"iframeTarget\\"></iframe>


<div>
============================
EXPORTS:
============================
{
	\\"default\\": {}
}
============================
</div></body></html>"
`;

exports[`"esModule" option should work when not specified and when the "injectType" option is "linkTag": errors 1`] = `Array []`;

exports[`"esModule" option should work when not specified and when the "injectType" option is "linkTag": warnings 1`] = `Array []`;

exports[`"esModule" option should work when not specified and when the "injectType" option is "singletonStyleTag": DOM 1`] = `
"<!DOCTYPE html><html><head>
  <title>style-loader test</title>
  <style id=\\"existing-style\\">.existing { color: yellow }</style>
<style>.foo {
    color: red;
}

.bar {
    color: blue;
}

.baz {
    color: coral;
}
</style></head>
<body>
  <h1>Body</h1>
  <div class=\\"target\\"></div>
  <iframe class=\\"iframeTarget\\"></iframe>


<div>
============================
EXPORTS:
============================
{
	\\"default\\": {}
}
============================
</div></body></html>"
`;

exports[`"esModule" option should work when not specified and when the "injectType" option is "singletonStyleTag": errors 1`] = `Array []`;

exports[`"esModule" option should work when not specified and when the "injectType" option is "singletonStyleTag": warnings 1`] = `Array []`;

exports[`"esModule" option should work when not specified and when the "injectType" option is "styleTag": DOM 1`] = `
"<!DOCTYPE html><html><head>
  <title>style-loader test</title>
  <style id=\\"existing-style\\">.existing { color: yellow }</style>
<style>.foo {
    color: red;
}

.bar {
    color: blue;
}

.baz {
    color: coral;
}
</style></head>
<body>
  <h1>Body</h1>
  <div class=\\"target\\"></div>
  <iframe class=\\"iframeTarget\\"></iframe>


<div>
============================
EXPORTS:
============================
{
	\\"default\\": {}
}
============================
</div></body></html>"
`;

exports[`"esModule" option should work when not specified and when the "injectType" option is "styleTag": errors 1`] = `Array []`;

exports[`"esModule" option should work when not specified and when the "injectType" option is "styleTag": warnings 1`] = `Array []`;

exports[`"esModule" option should work with a value equal to "false" and when the "injectType" option is "lazySingletonStyleTag": DOM 1`] = `
"<!DOCTYPE html><html><head>
  <title>style-loader test</title>
  <style id=\\"existing-style\\">.existing { color: yellow }</style>
<style>.foo {
    color: red;
}

.bar {
    color: blue;
}

.baz {
    color: coral;
}
</style></head>
<body>
  <h1>Body</h1>
  <div class=\\"target\\"></div>
  <iframe class=\\"iframeTarget\\"></iframe>


<div>
EXPORTED defaultImport:
============================
{}
============================
EXPORTED allImports:
============================
{
	\\"default\\": {}
}
============================
</div></body></html>"
`;

exports[`"esModule" option should work with a value equal to "false" and when the "injectType" option is "lazySingletonStyleTag": errors 1`] = `Array []`;

exports[`"esModule" option should work with a value equal to "false" and when the "injectType" option is "lazySingletonStyleTag": warnings 1`] = `Array []`;

exports[`"esModule" option should work with a value equal to "false" and when the "injectType" option is "lazyStyleTag": DOM 1`] = `
"<!DOCTYPE html><html><head>
  <title>style-loader test</title>
  <style id=\\"existing-style\\">.existing { color: yellow }</style>
<style>.foo {
    color: red;
}

.bar {
    color: blue;
}

.baz {
    color: coral;
}
</style></head>
<body>
  <h1>Body</h1>
  <div class=\\"target\\"></div>
  <iframe class=\\"iframeTarget\\"></iframe>


<div>
EXPORTED defaultImport:
============================
{}
============================
EXPORTED allImports:
============================
{
	\\"default\\": {}
}
============================
</div></body></html>"
`;

exports[`"esModule" option should work with a value equal to "false" and when the "injectType" option is "lazyStyleTag": errors 1`] = `Array []`;

exports[`"esModule" option should work with a value equal to "false" and when the "injectType" option is "lazyStyleTag": warnings 1`] = `Array []`;

exports[`"esModule" option should work with a value equal to "false" and when the "injectType" option is "linkTag": DOM 1`] = `
"<!DOCTYPE html><html><head>
  <title>style-loader test</title>
  <style id=\\"existing-style\\">.existing { color: yellow }</style>
<link rel=\\"stylesheet\\" href=\\"es-modules.css\\"></head>
<body>
  <h1>Body</h1>
  <div class=\\"target\\"></div>
  <iframe class=\\"iframeTarget\\"></iframe>


<div>
============================
EXPORTS:
============================
{
	\\"default\\": {}
}
============================
</div></body></html>"
`;

exports[`"esModule" option should work with a value equal to "false" and when the "injectType" option is "linkTag": errors 1`] = `Array []`;

exports[`"esModule" option should work with a value equal to "false" and when the "injectType" option is "linkTag": warnings 1`] = `Array []`;

exports[`"esModule" option should work with a value equal to "false" and when the "injectType" option is "singletonStyleTag": DOM 1`] = `
"<!DOCTYPE html><html><head>
  <title>style-loader test</title>
  <style id=\\"existing-style\\">.existing { color: yellow }</style>
<style>.foo {
    color: red;
}

.bar {
    color: blue;
}

.baz {
    color: coral;
}
</style></head>
<body>
  <h1>Body</h1>
  <div class=\\"target\\"></div>
  <iframe class=\\"iframeTarget\\"></iframe>


<div>
============================
EXPORTS:
============================
{
	\\"default\\": {}
}
============================
</div></body></html>"
`;

exports[`"esModule" option should work with a value equal to "false" and when the "injectType" option is "singletonStyleTag": errors 1`] = `Array []`;

exports[`"esModule" option should work with a value equal to "false" and when the "injectType" option is "singletonStyleTag": warnings 1`] = `Array []`;

exports[`"esModule" option should work with a value equal to "false" and when the "injectType" option is "styleTag": DOM 1`] = `
"<!DOCTYPE html><html><head>
  <title>style-loader test</title>
  <style id=\\"existing-style\\">.existing { color: yellow }</style>
<style>.foo {
    color: red;
}

.bar {
    color: blue;
}

.baz {
    color: coral;
}
</style></head>
<body>
  <h1>Body</h1>
  <div class=\\"target\\"></div>
  <iframe class=\\"iframeTarget\\"></iframe>


<div>
============================
EXPORTS:
============================
{
	\\"default\\": {}
}
============================
</div></body></html>"
`;

exports[`"esModule" option should work with a value equal to "false" and when the "injectType" option is "styleTag": errors 1`] = `Array []`;

exports[`"esModule" option should work with a value equal to "false" and when the "injectType" option is "styleTag": warnings 1`] = `Array []`;

exports[`"esModule" option should work with a value equal to "true" and when the "injectType" option is "lazySingletonStyleTag": DOM 1`] = `
"<!DOCTYPE html><html><head>
  <title>style-loader test</title>
  <style id=\\"existing-style\\">.existing { color: yellow }</style>
<style>.foo {
    color: red;
}

.bar {
    color: blue;
}

.baz {
    color: coral;
}
</style></head>
<body>
  <h1>Body</h1>
  <div class=\\"target\\"></div>
  <iframe class=\\"iframeTarget\\"></iframe>


<div>
EXPORTED defaultImport:
============================
{}
============================
EXPORTED allImports:
============================
{
	\\"default\\": {}
}
============================
</div></body></html>"
`;

exports[`"esModule" option should work with a value equal to "true" and when the "injectType" option is "lazySingletonStyleTag": errors 1`] = `Array []`;

exports[`"esModule" option should work with a value equal to "true" and when the "injectType" option is "lazySingletonStyleTag": warnings 1`] = `Array []`;

exports[`"esModule" option should work with a value equal to "true" and when the "injectType" option is "lazyStyleTag": DOM 1`] = `
"<!DOCTYPE html><html><head>
  <title>style-loader test</title>
  <style id=\\"existing-style\\">.existing { color: yellow }</style>
<style>.foo {
    color: red;
}

.bar {
    color: blue;
}

.baz {
    color: coral;
}
</style></head>
<body>
  <h1>Body</h1>
  <div class=\\"target\\"></div>
  <iframe class=\\"iframeTarget\\"></iframe>


<div>
EXPORTED defaultImport:
============================
{}
============================
EXPORTED allImports:
============================
{
	\\"default\\": {}
}
============================
</div></body></html>"
`;

exports[`"esModule" option should work with a value equal to "true" and when the "injectType" option is "lazyStyleTag": errors 1`] = `Array []`;

exports[`"esModule" option should work with a value equal to "true" and when the "injectType" option is "lazyStyleTag": warnings 1`] = `Array []`;

exports[`"esModule" option should work with a value equal to "true" and when the "injectType" option is "linkTag": DOM 1`] = `
"<!DOCTYPE html><html><head>
  <title>style-loader test</title>
  <style id=\\"existing-style\\">.existing { color: yellow }</style>
<link rel=\\"stylesheet\\" href=\\"es-modules.css\\"></head>
<body>
  <h1>Body</h1>
  <div class=\\"target\\"></div>
  <iframe class=\\"iframeTarget\\"></iframe>


<div>
============================
EXPORTS:
============================
{
	\\"default\\": {}
}
============================
</div></body></html>"
`;

exports[`"esModule" option should work with a value equal to "true" and when the "injectType" option is "linkTag": errors 1`] = `Array []`;

exports[`"esModule" option should work with a value equal to "true" and when the "injectType" option is "linkTag": warnings 1`] = `Array []`;

exports[`"esModule" option should work with a value equal to "true" and when the "injectType" option is "singletonStyleTag": DOM 1`] = `
"<!DOCTYPE html><html><head>
  <title>style-loader test</title>
  <style id=\\"existing-style\\">.existing { color: yellow }</style>
<style>.foo {
    color: red;
}

.bar {
    color: blue;
}

.baz {
    color: coral;
}
</style></head>
<body>
  <h1>Body</h1>
  <div class=\\"target\\"></div>
  <iframe class=\\"iframeTarget\\"></iframe>


<div>
============================
EXPORTS:
============================
{
	\\"default\\": {}
}
============================
</div></body></html>"
`;

exports[`"esModule" option should work with a value equal to "true" and when the "injectType" option is "singletonStyleTag": errors 1`] = `Array []`;

exports[`"esModule" option should work with a value equal to "true" and when the "injectType" option is "singletonStyleTag": warnings 1`] = `Array []`;

exports[`"esModule" option should work with a value equal to "true" and when the "injectType" option is "styleTag": DOM 1`] = `
"<!DOCTYPE html><html><head>
  <title>style-loader test</title>
  <style id=\\"existing-style\\">.existing { color: yellow }</style>
<style>.foo {
    color: red;
}

.bar {
    color: blue;
}

.baz {
    color: coral;
}
</style></head>
<body>
  <h1>Body</h1>
  <div class=\\"target\\"></div>
  <iframe class=\\"iframeTarget\\"></iframe>


<div>
============================
EXPORTS:
============================
{
	\\"default\\": {}
}
============================
</div></body></html>"
`;

exports[`"esModule" option should work with a value equal to "true" and when the "injectType" option is "styleTag": errors 1`] = `Array []`;

exports[`"esModule" option should work with a value equal to "true" and when the "injectType" option is "styleTag": warnings 1`] = `Array []`;
